cmd/apiserver/app/server/server.go

```diff
@@ -57,151 +36,81 @@ const (
+		return nil
 	}
-
-	// Store resources in etcd under our special prefix
-	options.EtcdOptions.StorageConfig.Prefix = etcdPathPrefix
+	glog.V(4).Infof("using storage type %s", storageTypeString)
 
 	// Create the command that runs the API server
 	cmd := &cobra.Command{
 		Short: "run a service-catalog server",
-		Run: func(c *cobra.Command, args []string) {
-			options.runServer()
-		},
 	}
-
 	// We pass flags object to sub option structs to have them configure
 	// themselves. Each options adds its own command line flags
 	// in addition to the flags that are defined above.
 	flags := cmd.Flags()
-	// TODO consider an AddFlags() method on our options
-	// struct. Will need to import pflag.
-	//
-	// repeated pattern seems like it should be refactored if all
-	// options were of an interface type that specified AddFlags.
 	flags.AddGoFlagSet(flag.CommandLine)
-	options.GenericServerRunOptions.AddUniversalFlags(flags)
-	options.SecureServingOptions.AddFlags(flags)
-	options.EtcdOptions.AddFlags(flags)
-	options.AuthenticationOptions.AddFlags(flags)
-	options.AuthorizationOptions.AddFlags(flags)
-
-	return cmd
-}
-
-// runServer is a method on the options for composition. allows embedding in a
-// higher level options as we do the etcd and serving options.
-func (serverOptions ServiceCatalogServerOptions) runServer() error {
-	glog.V(4).Infoln("Preparing to run API server")
-	// options
-	// runtime options
-	if err := serverOptions.GenericServerRunOptions.DefaultExternalAddress(serverOptions.SecureServingOptions, nil); err != nil {
-		return err
-	}
-
-	// server configuration options
-	glog.V(4).Infoln("Setting up secure serving options")
-	if err := serverOptions.SecureServingOptions.MaybeDefaultWithSelfSignedCerts(serverOptions.GenericServerRunOptions.AdvertiseAddress.String()); err != nil {
-		glog.Errorf("Error creating self-signed certificates: %v", err)
-		return err
-	}
-
-	// etcd options
-	if errs := serverOptions.EtcdOptions.Validate(); len(errs) > 0 {
-		glog.Errorln("Error validating etcd options, do you have `--etcd-servers localhost` set?")
-		return errs[0]
-	}
 
-	// config
-	glog.V(4).Infoln("Configuring generic API server")
-	genericconfig := genericapiserver.NewConfig().ApplyOptions(serverOptions.GenericServerRunOptions)
-	// these are all mutators of each specific suboption in serverOptions object.
-	// this repeated pattern seems like we could refactor
-	if _, err := genericconfig.ApplySecureServingOptions(serverOptions.SecureServingOptions); err != nil {
-		glog.Errorln(err)
-		return err
+	baseOptions := &ServiceCatalogServerBaseOptions{
+		GenericServerRunOptions: genericserveroptions.NewServerRunOptions(),
+		SecureServingOptions:    genericserveroptions.NewSecureServingOptions(),
+		AuthenticationOptions:   genericserveroptions.NewDelegatingAuthenticationOptions(),
+		AuthorizationOptions:    genericserveroptions.NewDelegatingAuthorizationOptions(),
 	}
```

Comment:
@krancour The way I did it wasn't much for memory efficiency, instead it was because I thought it would make code more readable. If you think this alternative is better, I'm good with it.